src/config.rs

```diff
@@ -28,6 +27,63 @@ pub enum ConfigError {
     AmbiguousSource(PathBuf, PathBuf),
 }
 
+/// Set of configs which can be merged as needed.
+///
+/// Sources from the lowest precedence:
+/// 1. Default
+/// 2. Base environment variables
+/// 3. User config `~/.jjconfig.toml` or `$JJ_CONFIG`
+/// 4. TODO: Repo config `.jj/repo/config.toml`
+/// 5. TODO: Workspace config `.jj/config.toml`
+/// 6. Override environment variables
+/// 7. TODO: Command-line arguments `--config-toml`
+#[derive(Clone, Debug)]
+pub struct LayeredConfigs {
+    default: config::Config,
+    env_base: config::Config,
+    user: Option<config::Config>,
+    env_overrides: config::Config,
+}
```

Comment:
Nice, this means that we'll be able to implement `jj config list --source` without depending on support for that in the `config` crate.

Reply:
Yes, that's the reason not to merge `default` and `env_base` upfront, but we might have to split `user` since it may be loaded from a directory.